Refactor scrypt hashing and source app session mgmt pt 1/3

nabla-c0d3 opened the pull request:
## Status

Ready.

## Description of Changes

This PR started with the goal of refactoring **CryptoUtil.hash_codename()** to move it out of CryptoUtil (for #5599).

However, I ran into some problems as I was trying to do this, and saw some opportunities for larger improvements. The downside is that the changes are bigger, so I split them into 3 PRs so they're a bit easier to review. While I know big changes are annoying to review, merge and ultimately deal with, I think these ones are important and worth it.

Overall I felt that a crucial concept was "missing"/invisible in the code base: a user who is currently logged into the source application. 

So the main change of these PRs is the introduction of a **SourceUser** class, which ties it all together: how a user's session is managed (when they create an account, login, etc.), how the user's crypto works (filesystem_id, gpg_secret, etc.), and the user's corresponding "Source" record in the DB. 

My overall goal is to make all this logic self-documenting so it is clear to the reader how these users (and more generally the source app) work. Currently, this logic is spread throughout the code base and also repeated a lot, which makes it hard to understand and hard to change. The main example of this is how hash_codename() is used in a lot of code locations, but in each of them it is hard to understand what is getting hashed, and why.

While I split the changes into 3 PRs, each PR brings something useful independently of the other ones:

* 1/3: This PR: Implement SourceUser and create_source_user() to consolidate source creation (More details below)
* 2/3: https://github.com/freedomofpress/securedrop/pull/5694 Remove CryptoUtil.hash_codename(), scrypt Python dependency, and refactor session management in the source app.
* 3/3: https://github.com/freedomofpress/securedrop/pull/5695 Remove usage of flask.g in the source app and fix tests.

If everything gets merged, the following issues will be completed or will see progress: #5455, #5599, #5613. I also think the source app's code base will be significantly easier to approach, understand, and maintain (at least that's the main goal).

**About this PR which is part 1/3:**

* Implement SourceUser and create_source_user() to consolidate source creation. In the current code base there is some duplicate code (in the tests, the apps, etc.) for creating a new source. This PR replaces all of them with a call to create_source_user() so there's one consistent way of creating a source.
* Fixes https://github.com/freedomofpress/securedrop/issues/5455.

nabla-c0d3 commented on securedrop/crypto_util.py:
```diff
@@ -162,33 +165,14 @@ def hash_codename(self, codename: DicewarePassphrase, salt: Optional[str] = None
-            passphrase=hashed_secret,
-            name_email=name,
+            passphrase=source_user.gpg_secret,
+            name_email=source_user.filesystem_id,
```
This makes the logic self-documenting and the docstring is no longer needed.

nabla-c0d3 commented on securedrop/models.py:
```diff
@@ -455,10 +456,17 @@ def __repr__(self) -> str:
+            p=1,
+            backend=backend,
+        )
+        return scrypt_instance.derive(password.encode("utf-8"))
```
Needed for https://github.com/freedomofpress/securedrop/issues/5613

nabla-c0d3 commented on securedrop/passphrases.py:
```diff
@@ -13,7 +13,7 @@
 DicewarePassphrase = NewType("DicewarePassphrase", str)
 
 
```
Only updated a variable's name in this file.

nabla-c0d3 commented on securedrop/source_app/main.py:
```diff
@@ -66,53 +71,45 @@ def select_logo() -> werkzeug.Response:
-                # Issue 2386: don't log in on duplicates
-                del session['codename']
+            except (SourcePassphraseCollisionError, SourceDesignationCollisionError) as e:
+                current_app.logger.error("Could not create a source: {}".format(e))
```
This consolidates the app's behavior when something goes wrong during account creation.

redshiftzero replied:
behavior change from the perspective of the user when `SourcePassphraseCollisionError` is raised since they will now get a flashed message, this is fine since the `codename` is no longer in the session when the exception is raised

nabla-c0d3 commented on securedrop/source_app/utils.py:
```diff
@@ -78,21 +62,30 @@ def async_genkey(crypto_util_: CryptoUtil,
+            filesystem_id=filesystem_id,
+            gpg_secret=crypto_util_.hash_codename(codename, salt=crypto_util_.scrypt_gpg_pepper)
+        )
+        crypto_util_.genkeypair(source_user)
```
genkeypair() now takes a source user.

nabla-c0d3 commented on securedrop/source_user.py:
```diff
@@ -0,0 +1,120 @@
+    return SourceUser(source_db_record, filesystem_id, gpg_secret)
+
+
+class _SourceScryptManager:
```
The scrypt hashing logic becomes private and an "implementation detail" of how the source app and users work.

nabla-c0d3 commented on securedrop/tests/test_source_user.py:
```diff
@@ -0,0 +1,89 @@
+        assert expected_filesystem_id == filesystem_id
+
+        expected_gpg_secret = "AWCRZVPA6YTQ2A3552LZJW3VO7L3ZONDFT6A6VPRGPGQQSNENRLA3EVRW4LZYNSUV5QIKNFZMJ2BMOVORG43ZETV5ZCRQKLJNOC2BXY="  # noqa: E501
+        assert expected_gpg_secret == gpg_secret
```
The test vectors in this test were generated using the previous / non-cryptography implementation of scrypt, to ensure that the new implementation returns the same hashes when given the same input.

redshiftzero replied:
sick

emkll commented:
Hi @nabla-c0d3 , thank you for your continued contributions to the project. We are currently preparing a release this week and next, but will get around to providing comments/review to this PR after that. Given the size and the areas of the code this PR is affecting, it may take us a while to review this series of proposed changes.

nabla-c0d3 commented:
> Given the size and the areas of the code this PR is affecting, it may take us a while to review this series of proposed changes.

@emkll No problem and makes sense!

eloquence commented:
Hi @nabla-c0d3, thanks a lot for taking on this complex refactoring work. I just want to manage expectations on review -- we're currently very focused on finishing Ubuntu 20.04 support, and these PRs touch a lot of key areas of the code, so it may still take us several weeks to complete a thorough review of these changes. That said, if you have any questions in the meantime, please don't hesitate to ping.

nabla-c0d3 commented:
No rush on my side

eloquence commented:
Thanks for rebasing @nabla-c0d3 -- we're just coming out of the 1.8.0 release cycle and should be able to take a closer look in the next few days, once tests are passing. (Let us know if we can help with that.)

nabla-c0d3 commented:
@eloquence Sounds good. 

The test that is failing doesn't seem to be related to my changes. The latest run on the develop branch failed in a similar fashion (different test but same error "Directory not empty": https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2119/workflows/c6371546-78d2-4050-8513-b33960083a7d/jobs/52269).

I will try rebasing again on the next commit that gets merged into develop, and hopefully the tests will pass,

nabla-c0d3 commented:
Only one test failed, but I don't think it's related to my changes. AFAIK I haven't touched apparmor or anything related to it.

redshiftzero commented on securedrop/loaddata.py:
```diff
@@ -254,16 +255,16 @@ def add_source() -> Tuple[Source, str]:
+        source_app_storage=current_app.storage,
+    )
+    source = source_user.get_db_record()
+    source.is_pending = False
```
`source.pending`

nabla-c0d3 replied:
Good catch

redshiftzero commented on securedrop/tests/test_source.py:
```diff
@@ -220,28 +219,24 @@ def test_generate(source_app):
+        resp = app.post(url_for('main.create'), data={'tab_id': tab_id}, follow_redirects=True)
+        assert resp.status_code == 200
+        assert 'codename' not in session
+        assert 'logged_in' not in session
```
note to other maintainers: this PR does preserve the duplicate tab behavior implemented in the test steps in #5075

redshiftzero commented on securedrop/tests/test_source_user.py:
```diff
@@ -0,0 +1,89 @@
+
+        # When deriving the passphrase's filesystem ID and GPG secret
+        scrypt_mgr = _SourceScryptManager(
+            salt_for_gpg_secret="YrPAwKMyWN66Y2WNSt+FS1KwfysMHwPISG0wmpb717k=".encode(),
```
nit: slight preference for an all caps constant for clarity at the top of the file or at the top of the function, e.g. `TEST_SALT_GPG_SECRET` for the values for `salt_for_gpg_secret` and similar for `salt_for_filesystem_id`

redshiftzero commented on securedrop/loaddata.py:
```diff
@@ -254,16 +255,16 @@ def add_source() -> Tuple[Source, str]:
-    os.mkdir(current_app.storage.path(source.filesystem_id))
-
     # Generate source key
     current_app.crypto_util.genkeypair(source.filesystem_id, codename)
```
should be taking a `SourceUser` instance now?

nabla-c0d3 replied:
Correct - good catch

nabla-c0d3 replied:
Not detected by mypy due to https://github.com/freedomofpress/securedrop/issues/5599

redshiftzero reviewed (commented):
hey thanks for this! I definitely agree the `SourceUser` object makes the logic clearer. dropped a few notes inline, nits are optional

eloquence commented:
@nabla-c0d3 We'll aim to take another pass through this PR in the next two weeks (@kushaldas has offered to be point) - we'll probably not be able to merge it yet, as we'll ideally merge all 3 PRs together. Thanks again for all your work on this.

nabla-c0d3 commented:
@eloquence Sounds good. There is one test failing when running `make testinfra` (ie. not the app's code) but I haven't been able get `testinfra` to run at all on a fresh VM.

eloquence commented:
@nabla-c0d3 If you have cycles, would you mind rebasing your branch against `develop` once more? Major code removal just landed which will also ensure that tests pass for future changes. Thanks, and sorry this is taking a while - big changes :)

nabla-c0d3 commented:
> 
> 
> @nabla-c0d3 If you have cycles, would you mind rebasing your branch against `develop` once more? Major code removal just landed which will also ensure that tests pass for future changes. Thanks, and sorry this is taking a while - big changes :)

No problem - I just did the rebase and it went smoothly. There are two tests failing in testinfra now, but as mentioned before I have not been able to setup a VM that can run `make testinfra`.

nabla-c0d3 commented:
I have rebased this PR, but the same `test_aa_no_denies_in_syslog()` is still failing.

I haven't been able to figure it out because the CircleCI log does not show the full output (`Full output truncated (4869 lines hidden), use '-vv' to show`). I also haven't been able to get `make testinfra` to work in an Ubuntu VM - all the tests in testinfra just fail right away.
Being able to see the full output of the test would already help a lot, but I am not sure how to get to that.

zenmonkeykstop commented:
Hi @nabla-c0d3 - looks like a file `securedrop/source_user.py` was added - for staging and prod environments, new files need to be explicitly included in the apache2 apparmor profile at `install_files/ansible-base/roles/build-securedrop-app-code-deb-pkg/files/usr.sbin.apache2`. Not sure what's going on with `make testinfra`, will look into it, but hopefully CI should get a little further now.

(Also rebased again to pull in an unrelated CI fix)

nabla-c0d3 commented:
> for staging and prod environments, new files need to be explicitly included in the apache2 apparmor profile

Thanks for looking into this. Looks like CI is now passing! 🎉

kushaldas commented on securedrop/tests/test_manage.py:
```diff
@@ -224,7 +224,13 @@ def test_were_there_submissions_today(source_app, config):
         args = argparse.Namespace(data_root=data_root, verbose=logging.DEBUG)
 
         count_file = os.path.join(data_root, 'submissions_today.txt')
-        source, codename = db_helper.init_source_without_keypair()
```
`codename` was not used, so this is fine.

kushaldas reviewed (dismissed):
There is one comment from @redshiftzero https://github.com/freedomofpress/securedrop/pull/5692/files#r601688777 which can be addressed in this while doing the conflict fix. After that we can merge this.

To test 
- [x] I created a staging instance from `develop` and created users.
- [x] created sources on that
- [x] build debian package from this branch and redeployed
- [x] old users can login, new users can be created and login properly

The code changes look good to me. @nabla-c0d3 after you update the branch, I will do one quick round of check and merge. For now moving to the next PR. Thank you once again for doing this.

nabla-c0d3 committed:
Implement create_source_user() to consolidate source creation

Fix test

Fix tests

added securedrop/source_user.py to Apache Apparmor profile

minor fix to dev data loader

nabla-c0d3 commented:
@kushaldas Thanks for looking into this and testing it.

> There is one comment from @redshiftzero https://github.com/freedomofpress/securedrop/pull/5692/files#r601688777 which can be addressed in this while doing the conflict fix. After that we can merge this.

Ok I have made the change and fixed the conflict.

nabla-c0d3 commented on securedrop/tests/test_source_user.py:
```diff
@@ -0,0 +1,93 @@
+
+
+TEST_SALT_GPG_SECRET = "YrPAwKMyWN66Y2WNSt+FS1KwfysMHwPISG0wmpb717k="
+TEST_SALT_FOR_FILESYSTEM_ID = "mEFXIwvxoBqjyxc/JypLdvgMRNRjApoaM0OBNrxJM2E="
```
Moved these to the top of the file

zenmonkeykstop commented on securedrop/source_user.py:
```diff
@@ -0,0 +1,120 @@
+    except IntegrityError:
+        db_session.rollback()
+        raise SourcePassphraseCollisionError(
+            "Passphrase {} is already used by another Source".format(source_passphrase)
```
Though the chance of this being hit is vanishingly small (and logging is disabled by default for the SI), IMO passphrases of sources shouldn't be logged.

nabla-c0d3 replied:
Yeah I agree. I removed the passphrase from the exception's message and put the filesystem_id instead.

zenmonkeykstop reviewed (changes requested):
One minor point re: exception error messages
